src/tools/rbd_nbd/rbd-nbd.cc

```diff
@@ -89,6 +90,15 @@ static bool exclusive = false;
 #endif
 #define htonll(a) ntohll(a)
 
+static int do_unmap();
+
+static void handle_signal(int signum)
+{
+  assert(signum == SIGINT || signum == SIGTERM);
+  derr << "*** Got signal " << sig_str(signum) << " ***" << dendl;
+  do_unmap();
```

Comment:
It works, though unnecessary opens/close nbd device. This might be also an additional source of races, when the device is already closed by this rbd-nbd process and opened by another.

I think just doing `ioctl(nbd, NBD_DISCONNECT)` is enough. For this the signal handler need to know `nbd` though. You could just make `nbd` global static (initialized to -1 for safety) or use C++11 lambda function, something like below:
```
std::function<void()> handle_signal_callback;
static void handle_signal(int signum) {
   ...
  handle_signal_callback();
}

...
int do_map(int argc, const char *argv[]) {
  ....
  handle_signal_callback = [nbd]() { ioctl(nbd, NBD_DISCONNECT); };
  register_async_signal_handler_oneshot(SIGINT, handle_signal);
  ...
```